feat: implement mobile fidget drag-and-drop system and fix homebase feed

Jhonattan2121 opened the pull request:
# Development Report - Mobile Drag-and-Drop System

## 🎯 **What Was Done**
Fix the drag-and-drop system for fidgets in mobile preview and implement the immutable feed for homebase.

## 🔥 **Main Problems Solved**

### **1. Drag-and-Drop Wasn't Working**
**Problem**: Dragging fidgets in mobile editor didn't update the footer  
**Cause**: `Space.tsx` passed `Object.keys()` without sorting  
**Solution**: Sort by `mobileOrder` before rendering  
**Result**: ✅ Drag fully functional

### **2. Immutable Feed in Wrong Place**  
**Problem**: Feed appeared in all homebase tabs  
**Cause**: `isHomebasePath` too broad  
**Solution**: Changed to `pathname === '/homebase'`  
**Result**: ✅ Feed only in correct tab

### **3. Background Image Cut Off**
**Problem**: Mobile preview background incomplete  
**Cause**: `Image` component with issues  
**Solution**: CSS `background-image`  
**Result**: ✅ Complete visual

```

## Reasons to remove `MobileView`

- **Less code** – `MobileViewSimplified` has ~95 lines vs `MobileView` with 300+ lines  
- **Simpler** – Doesn’t reinvent the wheel; uses the existing layout system  
- **Fewer bugs** – Simpler code = less chance of errors  
- **Code reuse** – Leverages `LayoutFidgets["tabFullScreen"]` which already works  
- **Easier to maintain** – Changes to the layout system apply automatically  



## 📝 **Main Files Modified**

```
🔧 Critical Fix
└── Space.tsx                    # Sort by mobileOrder (MAIN FIX)

🎯 Immutable Feed  
├── ThemeSettingsEditor.tsx      # Feed logic + specific pathname

📱 Mobile Interface
├── MobileSettings.tsx           # Drag-and-drop interface
├── TabNavigation.tsx            # Footer with tabs
└── MobileViewSimplified.tsx     # Mobile container

🛠️ Utils and Support
├── layoutUtils.ts               # Sorting by mobileOrder
└── tabFullScreen/index.tsx      # Main layout
```

## 🔄 **How It Works Now**

```
1. 👤 User drags fidget in sidebar
2. 🔄 MobileSettings updates local order
3. 📤 Callback to ThemeSettingsEditor
4. 💾 Saves fidgetInstanceDatums with mobileOrder
5. ⭐ Space.tsx sorts by mobileOrder (KEY!)
6. 📱 MobileViewSimplified re-renders
7. ✅ TabNavigation shows new order
```

## 🐛 **Most Important Bugs Fixed**

| Bug | Location | What It Was | What It Became |
|-----|----------|-------------|----------------|
| **Drag with no effect** | `Space.tsx` | `Object.keys()` | `sort((a,b) => mobileOrder)` ⭐ |
| **Feed everywhere** | `ThemeSettingsEditor` | `startsWith('/homebase')` | `=== '/homebase'` |
| **Background cut off** | `Space.tsx` | `<Image fill>` | `background-image` CSS |

---


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Introduced a simplified mobile view component for improved mobile layout rendering.
  * Added explicit support for a contextual "Feed" tab on homebase paths, including custom icons and labels.
  * Enhanced tab navigation to always show the "Feed" tab with clear labeling and icon fallback.

* **Improvements**
  * Centralized and optimized mobile fidget ordering using explicit mobile layout configurations.
  * Modularized viewport, path, and configuration management with custom hooks for better maintainability.
  * Replaced complex mobile view with a streamlined rendering approach, improving performance and code clarity.
  * Added automatic creation and immutable management of the "Feed" fidget on homebase tabs.
  * Updated mobile preview styling with a modern phone frame overlay and refined container layout.
  * Improved drag-and-drop reconciliation by adjusting React keys for mini apps.
  * Refined theme and mini app state management with custom hooks for better update handling.

* **Bug Fixes**
  * Fixed mobile drag-and-drop ordering by enforcing deterministic mobile layout order.
  * Corrected feed visibility to appear only on exact homebase paths, preventing unintended display.

* **Style**
  * Enhanced mobile navbar and tab navigation styling for consistent z-index and fixed positioning.
  * Updated UI elements to reflect immutable mini apps with disabled interactions and distinct visual cues.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Jhonattan2121 committed:
feat: implement mobile fidget drag-and-drop system and fix homebase feed
- Fix mobile drag-and-drop: sort fidgets by mobileOrder in Space.tsx
- Implement contextual immutable feed for homebase main tab only
- Fix mobile preview background image rendering
- Add comprehensive mobile settings interface
- Update fidget ordering logic and state management

Jhonattan2121 committed:
fix: removal showFeedOnMobile={showFeedOnMobile} build fix

Jhonattan2121 committed:
fix: resolve 'Unknown' tab name for homebase immutable feed

- Handle special case for feed ID in getFidgetName functions
- Return 'Feed' when fidgetDatum doesn't exist in fidgetInstanceDatums
- Apply fix to both TabNavigation.tsx and MobileNavbar.tsx components

j-paterson commented on src/common/lib/theme/ThemeSettingsEditor.tsx:
```diff
@@ -89,29 +89,106 @@ export function ThemeSettingsEditor({
-        props?.mobileFidgetName ||
-        props?.fidgetName;
+  const pathname = usePathname();
+  const isHomebasseFeedTab = pathname === '/homebase';
```
/homebase/feed also is a homebase feed tab

j-paterson commented on src/app/(spaces)/Space.tsx:
```diff
@@ -93,30 +94,41 @@ export default function Space({
       return;
     }
 
-    const { cleanedLayout, cleanedFidgetInstanceDatums, hasChanges } = comprehensiveCleanup(
```
This cleanup does more than what you have here. You are only removing datums that aren't in the layout. 

Comprehensive cleanup does that, vice versa, and also checks for overlaps.

j-paterson reviewed (commented):
Storing the order of a fidget in each fidget setting is unnecessarily complex and doesn't follow the structure of the rest of the codebase.

This should be reworked to store the mobile order inside the layoutDetails.
<img width="352" height="544" alt="Screenshot 2025-07-11 at 7 07 34 PM" src="https://github.com/user-attachments/assets/acd19d3d-b717-4052-b65b-df6965facad2" />

j-paterson commented on src/app/(spaces)/Space.tsx:
```diff
@@ -150,131 +165,270 @@ export default function Space({
+                  width={344}
+                  height={744}
+                  className="pointer-events-none select-none absolute inset-0 z-10"
+                />
```
There's an SVG file for this in the MobilePreview component. Why did we switch back to not using that component?

Jhonattan2121 committed:
refactor: move mobile fidget order from individual settings to layoutDetails

- Remove mobileOrder property from FidgetSettingsStyle interface
- Add layouts.mobile.layout array to LayoutFidgetConfig for storing mobile order
- Update processTabFidgetIds to use mobile layout order from layoutDetails
- Refactor MobileViewSimplified to accept and use mobileLayoutOrder parameter
- Update Space.tsx to pass mobile layout order from layoutDetails
- Modify ThemeSettingsEditor to save mobile order in layout config instead of fidget settings

Jhonattan2121 committed:
refactor(MobileNavbar): remove duplicate code and optimize fidget handling

- Remove duplicated fidget icon logic in getTabIcon function
- Simplify getTabLabel function by unifying fidget data handling
- Eliminate redundant feed-specific logic that was already covered by general fidget handling
- Improve code maintainability and reduce bundle size

Jhonattan2121 committed:
fix: drag not working, returning to normal

Jhonattan2121 committed:
fix: fixed type - layouts do not exist in LayoutFidgetConfig

Jhonattan2121 committed:
fix: remove extra parameter from processTabFidgetIds call

- Remove incorrect layoutConfig.layouts?.mobile?.layout parameter
- Update useMemo dependencies to match function signature

Jhonattan2121 committed:
Merge branch 'canary' into feed-immutable-preview-mobile

Jhonattan2121 committed:
refactor: extract MobilePreview component and integrate with Space
- Refactored Space.tsx to use MobilePreview component instead of inline code
- Improved code organization and reusability

Jhonattan2121 committed:
fix: Use the internal PhoneFrame component

Jhonattan2121 committed:
fix: refactor mobile preview with PhoneFrame component
• Fix PhoneFrame.tsx white overlay issue by removing blocking white rectangle
• Replace Image component with PhoneFrame SVG for better customization
• Preserve theme functionality for both spaces and homebase

Jhonattan2121 committed:
fix: remove duplicate ThemeSettingsEditor portal in MobilePreview

Jhonattan2121 committed:
fix: visibility button working back

Jhonattan2121 committed:
fix: correct typo in isHomebasseFeedTab variable to isHomebaseFeedTab
- Rename isHomebasseFeedTab to isHomebaseFeedTab for consistency
- Remove the extra 's' to follow the correct 'homebase' convention

Jhonattan2121 committed:
fix: correct invalid z-index CSS class from z-level-2 to z-2

Jhonattan2121 committed:
fix: issue with zIndex in mobilePreview fixed

Jhonattan2121 committed:
refactor: improve layout and clean up ThemeSettingsEditor

- Optimized Space layout and improved mobile responsiveness
- Replaced manual cleanup with comprehensiveCleanup utility
- Added showFeedOnMobile prop and memoized layout props
- Integrated smart fidget ordering with extractFidgetIdsFromLayout
- Simplified ThemeSettingsEditor by removing unused dependencies
- Removed homebase-specific logic and redundant comments
- Improved code readability, maintainability, and performance

Jhonattan2121 committed:
feat: implement centralized mobile layout configuration

- Move mobile fidget ordering from individual settings to the Space layout configuration
- Add SpaceLayoutConfig with support for mobile/desktop/presentation modes
- Create utility functions for layout format conversion and migration
- Update preview components and mobile tabs to use centralized ordering
- Maintain backward compatibility with the old layout format
- Remove the dependency on individual fidget mobileOrder settings

Jhonattan2121 committed:
🎨 Centralize mobile layout configuration

Move fidget mobile ordering from individual settings to Space-level
configuration with multi-mode support and automatic migration.

Jhonattan2121 committed:
fix: resolve TypeScript errors in initial space configurations

Use getLayoutConfig utility instead of direct property access to ensure
compatibility with both layout formats.

Jhonattan2121 committed:
refactor: extend LayoutFidgetDetails and remove presentation layout

Jhonattan2121 committed:
refactor: extend LayoutFidgetDetails for multi-layout support

- Add layouts? property to LayoutFidgetDetails
- Remove presentation layout type
- Unify layout system to use only LayoutFidgetDetails
- Improve separation of concerns between desktop/mobile layouts

Jhonattan2121 committed:
refactor: complete layout system unification

All layout components now use the unified LayoutFidgetDetails structure with optional layouts property for desktop/mobile support.

Jhonattan2121 committed:
refactor: apply separation of concerns to layout components

- Optimize Space.tsx with focused helper functions
- Separate layout config logic in MobileViewSimplified.tsx
- Simplify desktop layout selection in DesktopView.tsx
- Split theme and mobile app management in ThemeSettingsEditor.tsx
- Improve performance by replacing heavy useMemo with direct calculations
- Maintain immutable feed functionality and +tab system

j-paterson reviewed (approved):
Could use some docs explaining the flow for switching to the multiple layouts system. Overall, great improvement from the original state of this PR.

Jhonattan2121 committed:
docs: add comprehensive documentation for multiple layouts system

- Add complete documentation suite covering multiple layouts implementation
- Include both Portuguese and English versions of all documentation
- Cover drag-and-drop system, layout migration, and homebase feed
- Document Separation of Concerns architecture and technical details
- Provide troubleshooting guides and extensibility information

Documentation includes:
- README with overview and navigation
- Multiple layouts system architecture
- Layout migration guide with backward compatibility
- Mobile drag-and-drop implementation details
- Immutable homebase feed system
- Complete changelog of technical changes

Jhonattan2121 committed:
Merge branch 'canary' into feed-immutable-preview-mobile
